seen or caught

JaskRendix opened the pull request:
It was a request from #1121 

Implementing a dictionary that allows to track the monsters seen or caught, and obviously unseen if there is no value. I opted for an int, because maybe in a future someone can expand the "levels". At the moment there is seen (1) and caught (2).

As usual the fix for add_monster and random_monster, included evolve_monster. Save_upgrader that automatically set caught (2) your monsters as well as in combat.py for a capture result. A couple of actions and one condition for modders.

kerizane commented on tuxemon/event/actions/add_monster.py:
```diff
@@ -76,3 +76,4 @@ def start(self) -> None:
         current_monster.current_hp = current_monster.hp
 
         trainer.add_monster(current_monster)
+        trainer.tuxepedia[monster_slug] = 2
```
Can we use a string enum?

```
class SeenStatus(str, Enum):
    seen = "seen"
    caught = "caught"
```

kerizane replied:
I am also considering counting the number of times you have seen / caught instead. There is a slight complication in that you might want to not count an NPC's monster multiple times, but we don't need to support this straight away.

What do you think?

JaskRendix replied:
I'm open to your proposal. I think enum is a good concept.

kerizane replied:
Let's go with counting then.

kerizane commented on tuxemon/event/conditions/tuxepedia_is.py:
```diff
@@ -0,0 +1,69 @@
+    Script parameters:
+        monster_slug: Monster slug name (e.g. "rockitten").
+        value: seen = 1 / caught = 2
+
```
As I said earlier, I think a string is better than a magic number.

JaskRendix commented:
I wasn't sure if below NPC or Monsters (party), so I did only this just to be sure. Then I'll update the other parts.

kerizane commented:
It doesn't need to be on the npc or the monster. You're just using this in the tuxepedia instead of an int.

JaskRendix commented:
Multiple times seen or caught would result in another class?

Update:
I forgot value, I'm updating.

JaskRendix commented:
I tried to put SeenStatus.caught instead of caught, but I got
{'rockitten': <SeenStatus.caught: 'caught'>}

kerizane commented:
`{'rockitten': <SeenStatus.caught: 'caught'>}`

This is fine, that's just what an enum looks like.

If you're tracking the number of times then you don't really need to use an enum.

```
'rockitten': {
'seen': 10,
'caught': 0,
}
 ```

You can use a typeddict to define the structure.

JaskRendix committed:
seen or caught

JaskRendix commented:
Thanks, I just added SeenStatus.*.value for simplifying, so it prints as "caught" or "seen"

I removed import os from random_monster because I forgot to delete it when I created the file. It's unused.

JaskRendix commented:
We can integrate the counting separately by counting both.
Think about this following draft:
```
class GenericCount(TypedDict):
    creature: str
    seen_count: int
    caught_count: int
    evolved_count: int


    def get_count(
        self,
        monster: Monster,
        seen: int = None,
        caught: int = None,
        evolved: int = None
    ) -> GenericCount:
        return {
            "creature": monster,
            "seen_count": +seen,
            "caught_count": +caught,
            "evolved_count": +evolved,
        }
```
It can become a sort of statistic for the monster itself.

kerizane commented:
Have you tested combat against a wild monster and against an NPC?

If you run `make validate` from the base of the repo are there any errors?

JaskRendix commented:
I tested both combats.
Against wild monster registers the monster as seen, if I capture it, then it becomes caught. If I meet again the monster (already caught), then it remains caught as should.
Against trainers, it registers all the monsters as seen (I fought Billie in route2) and it registers correctly the evolution too (rockitten to rockat).

Before the battle with Billie
![Screenshot from 2022-09-27 14-27-04](https://user-images.githubusercontent.com/64643719/192527735-7fd3b082-814e-4fae-82f5-0733303e42db.png)
After the battle with Billie
![Screenshot from 2022-09-27 14-28-38](https://user-images.githubusercontent.com/64643719/192527740-faf7ca3f-c40d-4cdb-91c4-b78b4a72a6f1.png)

make validate
```
PYTHONPATH=. python scripts/schema.py --validate
pygame 2.1.2 (SDL 2.0.16, Python 3.10.6)
Hello from the pygame community. https://www.pygame.org/contribute.html
```
and if I do "python scripts/schema.py --validate"
```
Traceback (most recent call last):
  File "/home/nik/Tuxemon/scripts/schema.py", line 5, in <module>
    from tuxemon.db import (
ModuleNotFoundError: No module named 'tuxemon'
```

JaskRendix commented:
I need to check a couple of things with the actions and condition

JaskRendix commented:
Checked actions and condition, it works.

kerizane commented:
Some minor changes I want, will explain when I have time.

kerizane commented on tuxemon/event/actions/add_monster.py:
```diff
@@ -76,3 +77,4 @@ def start(self) -> None:
         current_monster.current_hp = current_monster.hp
 
         trainer.add_monster(current_monster)
+        trainer.tuxepedia[monster_slug] = SeenStatus.caught.value
```
```suggestion
        trainer.tuxepedia[monster_slug] = SeenStatus.caught
```

kerizane replied:
I'd rather you didn't use value everywhere, because then the typehints that claim this is SeenStatus are wrong.

If you don't like the way this gets printed then you can modify the way SeenStatus prints things.

Add the following method to SeenStatus

```
def __str__(self):
    return self.value
```

Not tested, but I think it should work.

Just so you know, the behaviour for classes that inherit from str and enum is like so:

```
>SeenStatus.seen.value == "seen"
True
>SeenStatus.seen == "seen"
True
```

JaskRendix replied:
Thanks, you're right about value. I appreciate the feedback.

kerizane commented on tuxemon/save_upgrader.py:
```diff
@@ -83,6 +84,12 @@ def upgrade_save(save_data: Dict[str, Any]) -> SaveData:
+
+    # set as captured the party monsters
+    if not save_data["tuxepedia"]:
+        for mons in save_data.get("monsters", []):
```
Note we also have monster in storage.

JaskRendix replied:
You're right

kerizane commented on tuxemon/states/combat/combat.py:
```diff
@@ -379,6 +380,9 @@ def transition_phase(self, phase: CombatPhase) -> None:
                 var["battle_last_monster_category"] = monster_record.category
                 var["battle_last_monster_shape"] = monster_record.shape
+                # Avoid reset string to seen if monster has already been caught
+                if monster_record.slug not in self.players[0].tuxepedia.keys():
```
```suggestion
                if monster_record.slug not in self.players[0].tuxepedia:
```

kerizane commented on tuxemon/states/combat/combat.py:
```diff
@@ -977,6 +981,10 @@ def perform_action(
                     if result["success"]:
+                        # Tuxepedia: set monster as caught (2)
+                        self.players[0].tuxepedia[
+                            self.monsters_in_play[self.players[1]][0].slug
```
can we just use `target` rather than `self.monsters_in_play[self.players[1]][0].slug`  ?

JaskRendix replied:
ok

kerizane commented:
If you fix these I will merge. I don't want to overcomplicate it by counting the monsters.

Make sure you test your changes.

JaskRendix committed:
Update tuxemon/states/combat/combat.py

Co-authored-by: Keri Zane Volans <13522244+kerizane@users.noreply.github.com>

JaskRendix committed:
Update tuxemon/event/actions/add_monster.py

Co-authored-by: Keri Zane Volans <13522244+kerizane@users.noreply.github.com>

kerizane commented on tuxemon/event/actions/random_monster.py:
```diff
@@ -80,3 +79,4 @@ def start(self) -> None:
         current_monster.current_hp = current_monster.hp
 
         trainer.add_monster(current_monster)
+        trainer.tuxepedia[monster_slug] = SeenStatus.caught.value
```
`SeenStatus.caught`

please do this for all cases of SeenStatus

JaskRendix replied:
Exactly what I'm doing

JaskRendix commented:
Testing is ok, it lacks only monster_boxes in save_upgrader

kerizane commented:
> Testing is ok, it lacks only monster_boxes in save_upgrader

Great, add that and I will merge.

kerizane commented on tuxemon/event/conditions/tuxepedia_is.py:
```diff
@@ -0,0 +1,72 @@
+                if player.tuxepedia[monster_key] == SeenStatus.caught:
+                    return True
+            elif monster_str == "seen":
+                if player.tuxepedia[monster_key] == SeenStatus.seen:
```
Shouldn't we allow caught monsters to be classed as seen? You could add "seen_only" as an option.

kerizane replied:
Or do we leave it as is and add seen_or_caught?

Which is least surprising?

JaskRendix replied:
It makes sense what you say.

I saw caught as the upgrade of seen. I did this for the Tuxedex, so it can calculate how many seen and how many caught of the total number of monsters as well as for modders that can use for trigger an event.

kerizane replied:
So, don't we need to fix this?

If something has been caught then we should return true if we ask if is seen.

JaskRendix replied:
True, I'll update.

JaskRendix commented:
Kennel is the box, I guess, because with this:
```
        for monstervalue in save_data.get("monster_boxes"):
            save_data["tuxepedia"][monstervalue] = SeenStatus.caught
```
I got `Kennel': <SeenStatus.caught: 'caught'>`

kerizane commented:
Check the rest of the code, but monster_boxes is probably a dictionary of box names against lists of monsters.
```
        for monsters in save_data.get("monster_boxes", {}).values():             
```

JaskRendix committed:
seen or caught

JaskRendix commented:
Done, updated.

It could be useful, I printed the structure save_data["monsters"].items():
```
('Kennel',[{'current_hp': 85, 'level': 5, 'name': 'Cateye', 'slug': 'cateye', 'total_experience': 125, 'gender': 'female', 'instance_id': '69457433c83d49e8aab324ec2a34a791', 'moves': ['amnesia', 'eyebite', 'flow']},{another monster, etc..}])
```

JaskRendix committed:
seen true if caught
